### PR TITLE
Redact query from HttpTelemetry

### LINKED ISSUE #74340: Disabling/filtering query parameters for URIs in HTTP tracing
Similar to #68675, we should consider some kind of configuration for disabling/filtering query parameters for URIs in HTTP tracing.

We should also discuss whether it should be a single config for tracing and logging, or they should be different.

Last time we discussed it, we believed it will be better to keep them separate for discoverability -- configuring HttpClientFactory's logging would make more sense as a part of general HttpClientFactory configuration.

Consuming tracing requires more conscious effort, so a setting there might be less evident (e.g. AppContext switch).

We also discussed that we generally don't like messing with all Uris at once (e.g. setting that will apply globally to all Uri.ToString)

Note that Barry's opinion was to have a single setting:

> It’s difficult, because I can see folks taking trace events and saving them for debugging or security parsing. And once that happens it’s exactly the same problem. I’d say match the two.

So I wonder maybe tracing's setting can apply to logging, but not the other way around.

### PR DESCRIPTION
**Pull request by ManickaP**
Follow up on #103769

Redacts query string from the official HttpTelemetry (EventSource "System.Net.Http"). Uses the same opt out as distributed tracing (Activity) and `HttpClientFactory` logging. Doesn't touch private events (EventSource "Private.InternalDiagnostics.*").

Fixes #74340

cc @CarnaViire

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs [hunk 1]
```diff
@@ -37,6 +38,17 @@ public static class Keywords
...
             Interlocked.Increment(ref _startedRequests);
+            if (!GlobalHttpSettings.DiagnosticsHandler.DisableUriRedaction)
+            {
+                int queryIndex = pathAndQuery.IndexOf('?');
+                if (queryIndex >= 0)
+                {
+                    using ValueStringBuilder str = queryIndex < 254 ? new ValueStringBuilder(stackalloc char[queryIndex + 2]) : new ValueStringBuilder(queryIndex + 2);
+                    str.Append(pathAndQuery.AsSpan(0, queryIndex + 1));
+                    str.Append('*');
+                    pathAndQuery = str.ToString();
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:49]:
```suggestion
                    pathAndQuery = $"{pathAndQuery.AsSpan(0, queryIndex + 1)}*";
```

**Review by MihaZupan — APPROVED** [review]

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs (hunk 1 above)
Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs]: Fixed.

**Review by CarnaViire — COMMENTED** [review]
There's also

https://github.com/dotnet/runtime/blob/0f2e63fbdd63bc239df46efd01a329b42fea8531/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs#L58

Is there a reason why we're skipping it?

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs [hunk 2]
```diff
@@ -37,6 +38,14 @@ public static class Keywords
         private void RequestStart(string scheme, string host, int port, string pathAndQuery, byte versionMajor, byte versionMinor, HttpVersionPolicy versionPolicy)
         {
             Interlocked.Increment(ref _startedRequests);
+            if (!GlobalHttpSettings.DiagnosticsHandler.DisableUriRedaction)
+            {
+                int queryIndex = pathAndQuery.IndexOf('?');
+                if (queryIndex >= 0)
```
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:44]: This is different from what ILogger is doing ATM:

https://github.com/dotnet/runtime/blob/0f2e63fbdd63bc239df46efd01a329b42fea8531/src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs#L173-L174

I guess we need to align in one way or another?

Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs]: Fixed.

**Comment by liveans** [comment]
Breaking change doc is created: https://github.com/dotnet/docs/issues/42794

**Pull request merged**